text/0000-macros-1.1.md

```diff
@@ -0,0 +1,473 @@
+details as possible, allowing the compiler to flexibly change how it interfaces
+to macros.
+
+## Macros 1.1
+
+Ok, with the background knowledge of what procedural macros are today, let's
+take a look at how we can solve the major problems blocking its stabilization:
+
+* Sharing an API of the entire compiler
+* Frozen interface between the compiler and macros
+
+### `libmacro`
+
+Proposed in [RFC 1566](https://github.com/rust-lang/rfcs/pull/1566) and
+described in [this blog post](http://ncameron.org/blog/libmacro/) the
+distribution will now ship with a new `libmacro` crate available for macro
+authors. The intention here is that the gory details of how macros *actually*
+talk to the compiler is entirely contained within this one crate. The stable
+interface to the compiler is then entirely defined in this crate, and we can
+make it as small or large as we want. Additionally, like the standard library,
+it can contain unstable APIs to test out new pieces of functionality over time.
+
+The initial implementation of `libmacro` is proposed to be *incredibly* bare
+bones:
+
+```rust
+#![crate_name = "macro"]
+
+pub struct TokenStream {
+    // ...
+}
+
+#[derive(Debug)]
+pub struct FromSourceError {
+    // ...
+}
+
+pub struct MacroContext {
+    // ...
+}
+
+impl TokenStream {
+    pub fn from_source(source: &str) -> Result<TokenStream, FromSourceError> {
+        // ...
+    }
+
+    pub fn to_source(&self) -> String {
+        // ...
+    }
+}
+```
+
+That is, there will only be a handful of exposed types and `TokenStream` can
+only be converted to and from a `String`. Eventually `TokenStream` type will
+more closely resemble token streams [in the compiler
+itself][compiler-tokenstream], and more fine-grained manipulations will be
+available as well.
+
+Additionally, the `MacroContext` structure will initially be completely devoid
+of functionality, but in the future it will be the entry point for [many other
+features][macro20] one would expect in macros 2.0
+
+[compiler-tokenstream]: https://github.com/rust-lang/rust/blob/master/src/libsyntax/tokenstream.rs#L323-L338
+[macro20]: http://ncameron.org/blog/libmacro/
+
+### Defining a macro
+
+A new crate type will be added to the compiler, `rustc-macro`, indicating a
+crate that's compiled as a procedural macro. Like the executable, staticlib,
+and cdylib crate types the `rustc-macro` crate type is intended to be a final
+product.  What it *actually* produces is not specified, but if a `-L` path is
+provided to it then the compiler will recognize the output artifacts as a
+macro and it can be loaded for a program.
+
+Unlike today, however, there will not be a "registrar" function, but rather a
+number of functions which act as token stream tranformers to implement macro
+functionality.
+
+Putting this together, a macro crate might look like:
+
+```rust
+#![crate_type = "rustc-macro"]
+#![crate_name = "double"]
+
+extern crate macro;
```

Comment:
I think we need a better name for libmacro since `macro` is a keyword, although I guess we could make it contextual (I only want to use it for `macro!`). It is currently reserved.